src/runtime/objmodel.cpp

```diff
@@ -4002,28 +4064,165 @@ Box* typeNew(Box* _cls, Box* arg1, Box* arg2, Box** _args) {
+                if (may_add_weak && !add_weak)
+                    continue;
+                /* Nothing more to check */
+                break;
+            }
+        }
+    }
+
+    int attrs_offset = base->attrs_offset;
+    int dict_offset = base->tp_dictoffset;
+    int weaklist_offset = 0;
+    int basic_size = 0;
+
+    int cur_offset = base->tp_basicsize + sizeof(Box*) * final_slot_names.size();
+    if (add_dict) {
+        // CPython would set tp_dictoffset here, but we want to use attrs instead.
+        if (base->tp_itemsize) {
+            // A negative value indicates an offset from the end of the object
+            attrs_offset = -(long)sizeof(HCAttrs);
+        } else {
+            attrs_offset = cur_offset;
+        }
+        cur_offset += sizeof(HCAttrs);
     }
+    if (add_weak) {
+        assert(!base->tp_itemsize);
+        weaklist_offset = cur_offset;
+        cur_offset += sizeof(Box*);
+    }
+    basic_size = cur_offset;
+
+    size_t total_slots = final_slot_names.size()
+                         + (base->tp_flags & Py_TPFLAGS_HEAPTYPE ? static_cast<BoxedHeapClass*>(base)->nslots : 0);
+    BoxedHeapClass* made = BoxedHeapClass::create(metatype, base, NULL, attrs_offset, weaklist_offset, basic_size, true,
+                                                  name, bases, total_slots);
+    made->tp_dictoffset = dict_offset;
+
+    if (boxedSlots) {
+        // Set ht_slots
+        BoxedTuple* slotsTuple = BoxedTuple::create(final_slot_names.size());
+        for (size_t i = 0; i < final_slot_names.size(); i++)
+            (*slotsTuple)[i] = final_slot_names[i];
+        assert(made->tp_flags & Py_TPFLAGS_HEAPTYPE);
+        static_cast<BoxedHeapClass*>(made)->ht_slots = slotsTuple;
 
-    // TODO: how much of these should be in BoxedClass::finishInitialization()?
-    made->tp_dictoffset = base->tp_dictoffset;
+        BoxedHeapClass::SlotOffset* slot_offsets = made->slotOffsets();
+        size_t slot_offset_offset = made->tp_basicsize;
+
+        // Add the member descriptors
+        size_t offset = base->tp_basicsize;
+        for (size_t i = 0; i < final_slot_names.size(); i++) {
+            made->giveAttr(static_cast<BoxedString*>(slotsTuple->elts[i])->s.data(),
+                           new BoxedMemberDescriptor(BoxedMemberDescriptor::OBJECT_EX, offset, false /* read only */));
+            slot_offsets[i] = offset;
+            offset += sizeof(Box*);
+        }
+    }
+
+    // Add slot offsets for slots of the base
+    // NOTE: I don't think CPython does this
+    if (base->tp_flags & Py_TPFLAGS_HEAPTYPE) {
+        BoxedHeapClass::SlotOffset* slot_offsets = made->slotOffsets();
+        BoxedHeapClass* base_heap_cls = static_cast<BoxedHeapClass*>(base);
+        BoxedHeapClass::SlotOffset* base_slot_offsets = base_heap_cls->slotOffsets();
+        memcpy(&slot_offsets[final_slot_names.size()], base_slot_offsets,
+               base_heap_cls->nslots * sizeof(BoxedHeapClass::SlotOffset));
+    }
```

Comment:
ok spent some time trying to think about why we're doing this and cpython doesn't.  you should check what I'm saying and then add a comment if it's along the right lines :P

at first I thought CPython was directly constructing the member descriptors into this array.  it turns out that they just use the space to create PyMemberDef objects, which PyType_Ready will turn into normal descriptor objects.  It looks like they want to keep them around for the same reason we do, which is for GC traversing of the instances of the class.  except that they crawl the inheritance hierarchy (I guess you just need to check the main base?).

so are we just copying them all into here so that we can scan objects more quickly?  I think it's a decent tradeoff to make -- it increases the size of classes that inherit from classes with slots, but makes objects faster to visit with the GC.

also one thing I was confused about was I thought we were also adding member descriptors for the base slots, which would be user-visible.  so I think this copying is just for gc purposes?